src/monster.cpp

```diff
@@ -352,12 +353,13 @@ bool monster::can_drown() const
 
 bool monster::digging() const
 {
-    return has_flag(MF_DIGS) || (has_flag(MF_CAN_DIG) && g->m.has_flag("DIGGABLE", posx(), posy()));
+    return has_flag(MF_DIGS) || ( has_flag(MF_CAN_DIG) && underwater );
 }
 
 int monster::sight_range( const int light_level ) const
 {
-    if( !can_see() ) {
+    // Non-aquatic monsters can't see much when submerged
+    if( !can_see() || ( underwater && !has_flag( MF_SWIMS ) && !has_flag( MF_AQUATIC ) && !digging() ) ) {
```

Comment:
Doesn't have to happen *now* but nonaquatic critters also shouldn't be able to smell anything underwater.  (Aquatic critters can smell underwater just fine: see "shark".)